src/nvim/ex_cmds2.c

```diff
@@ -3014,6 +3014,44 @@ static FILE *fopen_noinh_readbin(char *filename)
   return fdopen(fd_tmp, READBIN);
 }
 
+typedef struct {
+  char_u *buf;
+  size_t offset;
+} GetStrLineCookie;
+
+static char_u *get_str_line(int c, void *cookie, int ident)
+{
+  GetStrLineCookie *p = cookie;
+  size_t i = p->offset;
+  if (strlen((char *)p->buf) <= p->offset) {
+    return NULL;
+  }
+  while (!(p->buf[i] == '\n' || p->buf[i] == '\0')) {
+    i++;
+  }
+  char buf[2046];
+  char *dst;
+  dst = xstpncpy(buf, (char *)p->buf+p->offset, i - p->offset);
+  if ((uint32_t)(dst - buf) != i - p->offset) {
+    smsg(_("nvim_source error parsing command %s"), p->buf);
+  }
+  buf[i-p->offset]='\0';
+  p->offset = i + 1;
+  return (char_u *)xstrdup(buf);
+}
+
+int do_source_str(char_u *cmd)
+{
+  int retval;
+  GetStrLineCookie cookie = {
+    .buf = cmd,
+    .offset = 0,
+  };
+  current_sctx.sc_sid = 999999;
```

Comment:
probably should use a defined constant. there already are some special values in `globals.h`, should probably add it after `SID_API_CLIENT` (and set it to -9).

Reply:
also I think we might need to save and restore `current_sctx` in case `nvim_source` is called by a script (unless it is automatically restored after every command, or something).

Reply:
I have saved and restored the script context and have also moved the script id to a constant. Do I need to also save `sourcing_name`, `sourcing_lnum` like I saw is being done in other places e.g. in `execute_env(..)` in `main.c`?

Reply:
I think so, `do_source` does it.